### ISSUE TITLE
Logging issue with 0.9.0 and current dev branch

### ISSUE THREAD

**Issue by ananyahjha93** [mosaicml/composer#1473]
** Environment **

- OS: Ubuntu 20.04
- Hardware (GPU, or instance type): 8xA100
- cuda: 11.3
- cudnn: 8
- pytorch: 1.12.1
- composer: dev branch installed from source/0.9.0 installed from pip
- transformers: 4.21.2

** To reproduce

I have the following definition of bloom model, mostly copied from the GPT2 definition within composer.

```
def create_bloom(
    model_name: str,
    tokenizer_name: str,
    use_pretrained: Optional[bool] = False,
    model_config: Optional[dict] = None,
    gradient_checkpointing: Optional[bool] = False,
) -> ComposerModel:

    if not model_config:
        model_config = {}

    if use_pretrained:
        model = transformers.AutoModelForCausalLM.from_pretrained(model_name, **model_config)
    else:
        config = transformers.AutoConfig.from_pretrained(model_name, **model_config)
        model = transformers.AutoModelForCausalLM.from_config(config)

    tokenizer = transformers.AutoTokenizer.from_pretrained(tokenizer_name)

    if gradient_checkpointing:
        model.gradient_checkpointing_enable()

    return HuggingFaceModel(model=model, tokenizer=tokenizer, metrics=[HFCrossEntropy(), Perplexity()])
```

There are 2 issues, one with the 0.9.0 release and the other with the dev branch.

1. Running LM training with grad accumulation with 0.9.0 doesn't plot HF metrics in wandb, but has correct step counts while logging metrics.

You can see that the logs don't show Perplexity and CrossEntropy metrics.
![image](https://user-images.githubusercontent.com/7491256/187063210-18da4fdd-381e-4255-af27-8ab55bf12d05.png)
![image](https://user-images.githubusercontent.com/7491256/187063229-3227b5d6-e9be-419c-8772-0eb8f303292b.png)

2. Running LM training with grad accumulation with the dev branch plots HF metrics but gets the step count while plotting these metrics completely wrong.

You can see metrics being plotted for 266 step with only 38 batches being trained.
![image](https://user-images.githubusercontent.com/7491256/187063267-d8be0b1d-4d5f-4e7c-a68e-cc3ebda9db5e.png)
![image](https://user-images.githubusercontent.com/7491256/187063280-921fe86e-d57a-4ec0-9424-fd697febc32d.png)

3. If I run the same training with deepspeed stage-2 enabled (dev branch), the metrics are plotted with correct step count.

## Expected behavior

Both Perplexity and CrossEntropy metrics are plotted with correct step count.

**Comment by ananyahjha93** [comment]
Update on point 2: Actually, I think there is a PR merged in the last couple of days which has introduced this issue. I had a dev branch installed on a separate machine, which is maybe 3 days old, which plots metric steps correctly, both with and without deepspeed.

**Comment by codestar12** [comment]
I noticed this as well. I'm going to dig into it a bit. Thanks!

**Comment by eracah** [comment]
Hey @ananyahjha93, it looks like the progress bar is counting the batches wrong, but weights and biases is plotting them correctly? Is that the issue?

**Comment by ananyahjha93** [comment]
@eracah no the progress bar is counting the batches correctly, because I am doing grad accumulation. Weights and biases is plotting the steps incorrectly.

**Comment by mvpatel2000** [comment]
> @eracah no the progress bar is counting the batches correctly, because I am doing grad accumulation. Weights and biases is plotting the steps incorrectly.

Just to be 100% clear, basically the issue is that the weights and bias x-axis is scaled by some factor that's not clear / doesn't sync with the progress bar. Is that correct?

**Comment by ananyahjha93** [comment]
yes

**Comment by mvpatel2000** [comment]
Great, thanks for the clarification. We'll look into this and get back to you. Thanks for raising this!

**Comment by ananyahjha93** [comment]
and this wasn't the case a few PR merges back because on a different server which has the dev branch from last week, this works fine.

**Comment by mvpatel2000** [comment]
Yep, makes sense. We're in the process of refactoring logging (and later checkpointing) to improve usability, and it looks like something weird got snuck in there.

**Comment by ananyahjha93** [comment]
thanks!

**Comment by eracah** [comment]
Hey, @ananyahjha93. @mvpatel2000 and I think we figured it out. It's very subtle, weird behavior by WandB.

So each of your two issues has a weird WandB reason behind it: 

1. If you log a metric and manually specify the timestamp and it's less than WandB's internal step count, then WandB won't log the metric. This is the case for your issue using 0.9. You can even check WandB's logs (usually located in `wandb/debug-internal.log` and you'll see something like `wandb: WARNING Step must only increase in log calls. Step 48 < 51; dropping {'metrics/eval/HFCrossEntropy': 5.3}`

2. However, if you don't specify step and let WandB handle the step automatically then it may increment the step after every log call. So if you log 3 other things before logging your HF metric, for example, then the metric will seemingly log every 4 steps. This seems to be the case for your issue using dev because we moved to automatic step logging in [PR 1416](https://github.com/mosaicml/composer/pull/1416)

I will try to push up a fix today using manual logging, but being extra careful to make sure the steps always increase (using `state.timestamp.batch` for all steps and not `state.eval_timestamp.batch` or `state.timestamp.batch_in_epoch`)

Let me know if that makes sense, @ananyahjha93. And thanks again for alerting us to this bug!

**Comment by eracah** [comment]
OK, @ananyahjha93, this pr that was just merged should fix the issue. Give it another try and let us know if it works!
